smithy-model/src/main/java/software/amazon/smithy/model/transform/ModelTransformer.java

```diff
@@ -491,4 +492,56 @@ public Model changeShapeType(Model model, Map<ShapeId, ShapeType> shapeToType) {
     public Model copyServiceErrorsToOperations(Model model, ServiceShape forService) {
         return new CopyServiceErrorsToOperationsTransform(forService).transform(this, model);
     }
+
+    /**
+     * Updates the model so that every operation has a dedicated input shape marked
+     * with the {@code input} trait and output shape marked with the {@code output}
+     * trait, and the targeted shapes all have a consistent shape name of
+     * OperationName + {@code inputSuffix} / {@code outputSuffix} depending on the
+     * context.
+     *
+     * <p>If an operation's input already targets a shape marked with the {@code input}
+     * trait, then the operation and input shape is left as-is regardless of its name.
```

Comment:
These conditionals still route through `renameShapeIfNeeded` in the transform ([input](https://github.com/awslabs/smithy/pull/992/files#diff-4bf748f6907e4f1457f9c288fd3e90fa9d01662d3fa84adebf3028f96aee36e5R107), [output](https://github.com/awslabs/smithy/pull/992/files#diff-4bf748f6907e4f1457f9c288fd3e90fa9d01662d3fa84adebf3028f96aee36e5R124)).